File: game/timer.py
"""
Timer module for Critter Bonk.
Manages the 30-second countdown, pause/resume time calculations, and elapsed time.
"""

import time
from typing import Optional, Callable


class GameTimer:
    DEFAULT_DURATION_SECONDS = 30.0

    def __init__(self, duration: float = DEFAULT_DURATION_SECONDS):
        self.total_duration = duration
        self.remaining_time = duration
        self.is_running = False
        self._last_tick_timestamp: Optional[float] = None
        self._on_tick_callback: Optional[Callable[[float], None]] = None
        self._on_complete_callback: Optional[Callable[[], None]] = None

    def start(self) -> None:
        self.is_running = True
        self._last_tick_timestamp = time.time()

    def pause(self) -> None:
        if not self.is_running:
            return
        self.update()
        self.is_running = False
        self._last_tick_timestamp = None

    def resume(self) -> None:
        if self.is_running or self.remaining_time <= 0:
            return
        self.is_running = True
        self._last_tick_timestamp = time.time()

    def update(self) -> float:
        """Updates the remaining time based on wall clock drift compensation."""
        if not self.is_running or self._last_tick_timestamp is None:
            return self.remaining_time

        now = time.time()
        delta = now - self._last_tick_timestamp
        self._last_tick_timestamp = now

        self.remaining_time = max(0.0, self.remaining_time - delta)
        if self._on_tick_callback:
            self._on_tick_callback(self.remaining_time)

        if self.remaining_time <= 0:
            self.is_running = False
            if self._on_complete_callback:
                self._on_complete_callback()

        return self.remaining_time

    @property
    def elapsed_time(self) -> float:
        return max(0.0, self.total_duration - self.remaining_time)

    def reset(self, duration: Optional[float] = None) -> None:
        if duration is not None:
            self.total_duration = duration
        self.remaining_time = self.total_duration
        self.is_running = False
        self._last_tick_timestamp = None

File: web/js/timer.js
/**
 * Critter Bonk - Timer Engineering Module
 * Drift-compensated countdown clock with strict interval safety, pause/resume, and zero leak guarantees.
 */

export class GameTimer {
  constructor(durationSeconds = 30.0) {
    this.totalDuration = durationSeconds;
    this.remainingTime = durationSeconds;
    this.isRunning = false;

    this.timerId = null;
    this.lastTickTime = null;

    this.onTick = null;       // callback(remainingTime, progressRatio)
    this.onComplete = null;   // callback()
  }

  /**
   * Starts or resumes the countdown timer.
   */
  start() {
    if (this.isRunning) {
      return;
    }

    this.isRunning = true;
    this.lastTickTime = performance.now();

    // Use a precise 50ms interval for ultra-smooth UI progress updates
    this.clearTimerLoop();
    this.timerId = window.setInterval(() => {
      this.tick();
    }, 50);

    // Initial immediate tick
    this.tick();
  }

  /**
   * Evaluates delta time with drift compensation.
   */
  tick() {
    if (!this.isRunning) return;

    const now = performance.now();
    const deltaMs = now - (this.lastTickTime || now);
    this.lastTickTime = now;

    this.remainingTime = Math.max(0.0, this.remainingTime - deltaMs / 1000.0);
    const progressRatio = Math.max(0.0, Math.min(1.0, this.remainingTime / this.totalDuration));

    if (typeof this.onTick === 'function') {
      this.onTick(this.remainingTime, progressRatio);
    }

    if (this.remainingTime <= 0.001) {
      this.stop();
      if (typeof this.onComplete === 'function') {
        this.onComplete();
      }
    }
  }

  /**
   * Pauses the timer, preserving remaining time without drift.
   */
  pause() {
    if (!this.isRunning) return;

    // Do a final tick to capture elapsed time right up to pause moment
    this.tick();
    this.isRunning = false;
    this.clearTimerLoop();
    this.lastTickTime = null;
  }

  /**
   * Resumes the timer from current remaining time.
   */
  resume() {
    if (this.isRunning || this.remainingTime <= 0) return;
    this.start();
  }

  /**
   * Stops and clears the timer.
   */
  stop() {
    this.isRunning = false;
    this.clearTimerLoop();
    this.lastTickTime = null;
  }

  /**
   * Resets timer to fresh duration.
   */
  reset(newDuration = null) {
    this.stop();
    if (newDuration !== null && newDuration > 0) {
      this.totalDuration = newDuration;
    }
    this.remainingTime = this.totalDuration;
  }

  clearTimerLoop() {
    if (this.timerId !== null) {
      window.clearInterval(this.timerId);
      this.timerId = null;
    }
  }
}
